Part One

lancylot2004 opened the pull request:
- Inputs for station and call to action for search.
- Redirect to LNER website upon action.
- Implementation of API singleton for station polling, ready for future expansion.
- [chore] Version bumps and minor gradle changes.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
feat: Initial setup.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
wip: Hook onto `MainActivity`.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
feat: Added API config.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
fix: Bumps `kotlinx-serialization` and related versions.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
feat: Adds `Client` singleton for server requests.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
feat: Implements part one UI.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
feat: Redirect to find live times.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
feat: Added unenabled button functionality and fixed lazy width issues.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 committed:
fix: Optimisations and minor improvements.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

lancylot2004 commented on app/src/main/kotlin/com/example/trainboard/api/Client.kt:
```diff
@@ -0,0 +1,52 @@
+    private val _stations = MutableStateFlow<Set<Station>>(emptySet())
+    val stations: StateFlow<Set<Station>> get() = _stations
+
+    init {
+        CoroutineScope(Dispatchers.IO).launch {
```
This should be fine...? Worth a look into and refactoring anyways, but since each coroutine scope invocation `CoroutineScope(...)` creates a new scope, none should affect the others...?

Discuss...

AmeliaIsCoding replied:
I suppose this could be problematic if it were possible to fetch the stations twice, starting the second call before the first one has resolved, such as if the fetch happened on clicking a button and the user did it twice in quick succession. If this were possible then we could make two calls with no way of knowing which to use. However because you're using a singleton I don't think this would ever happen, so I think your implemenation is fine. Do you agree with this?

lancylot2004 replied:
Agree that it would be if we fetched twice, but fortunately we don't :)

We'd probably want a better architecture anyways, away from the Singleton pattern, if things got more... concurrent :)

AmeliaIsCoding commented on app/build.gradle.kts:
```diff
@@ -1,38 +1,35 @@
-val apiKey = env["API_KEY"] ?: "default_api_key"
+val propertyFile = loadProperties("local.properties")
+val apiKey =
+    propertyFile.getProperty("trainboard.api_key")
```
looks neater but would be hard to extend if we added more environment variables so not a very robust solution!

I was hoping there might be some neat one-liner for env parsing in build.gradle.kts, but I compared with the LNER app approach and maybe there just isn't 🤷  happy for you to either keep your approach or revert to mine

lancylot2004 replied:
Could you please explain?

Another property would be like

```
val anotherProperty = propertyFile.getProperty("another.property")
```

And you just need in `local.properties`

```
another.property=something
```

AmeliaIsCoding replied:
Oops, I agree with you - my point was that indeed you'd have to add that line to `build.gradle.kts` everytime you added a build config value. I was saying that this isn't very maintainable. But I missed that in the initial solution you still have to add a line each time anyway. Happy for you to keep your solution :)

lancylot2004 replied:
Hmm, yes! I know of a few solutions that avoid any explicit code addition (except use-site of course), but those might be a bit overkill here... 

Will keep :)

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/api/Client.kt:
```diff
@@ -0,0 +1,52 @@
+                .let { it.stations.filter { station -> station.crs != null } }
+                .let { _stations.value = it.toSet() }
+        }
+    }
```
what happens if the API call fails? Can we handle this more gracefully?

lancylot2004 replied:
Yes we can! To come in Part 2... unless we should fix in part one too?

AmeliaIsCoding replied:
Yes I'd say add error handling at the point of adding any API call please :) If it causes major rebasing issues now if you've done it in part 2, feel free to just include in part 2 this time

NCHS6 replied:
Hello World!

NCHS6 replied:
Rebased and added simple error handling to the API to ignore errors instead of crashing. We have covered it more comprehensively in Part 2.

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/structures/Station.kt:
nice!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/structures/StationsResponse.kt:
nice!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+package com.example.trainboard.ui.theme
```
I agree with this CoPilot suggestion, do you?

AmeliaIsCoding replied:
Hi @lancylot2004 and Nick, did you see this comment please?

NCHS6 replied:
Implemented it!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun RootView(modifier: Modifier = Modifier) {
+    var fromStation by remember { mutableStateOf<Station?>(null) }
```
I would suggest that `departureStation` and `arrivalStation` are slightly more instantly understandable names

lancylot2004 replied:
> [!IMPORTANT]
> Modifications required.

AmeliaIsCoding replied:
Can't see any changes here :)

NCHS6 replied:
Updated!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+    label: String,
+    onStationChange: (Station) -> Unit,
+) {
+    var expanded by remember { mutableStateOf(false) }
```
`isExpanded` is a better variable name here (by Softwire standards) -> we typically use `is` prefixes for Booleans

lancylot2004 replied:
> [!IMPORTANT]
> Modifications required.

AmeliaIsCoding replied:
Can't see any changes here :)

NCHS6 replied:
Done!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+        expanded = expanded,
+        onExpandedChange = {
+            expanded = it
+            focusRequester.requestFocus()
```
nice!
what if expanded is being changed to `false`? do we still want to request focus?

lancylot2004 replied:
Good point!

lancylot2004 replied:
> [!IMPORTANT]
> Investigation required.

NCHS6 replied:
Tested and changed to only request focus if expanded

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+                .menuAnchor(MenuAnchorType.PrimaryEditable)
+                .fillMaxWidth()
+                .onFocusEvent {
+                    if (it.hasFocus) return@onFocusEvent
```
can you explain this line please? :)

AmeliaIsCoding replied:
@lancylot2004 and Nick, did you see this comment please?

lancylot2004 replied:
Which line?

AmeliaIsCoding replied:
`if (it.hasFocus) return@onFocusEvent` - the line that the comment is made on is always the bottom line!

lancylot2004 replied:
Good to know!

The code just below tries to detect input which corresponds to one of the stations in the list. (This is search bit that can be improved, e.g. with some fuzzy strategies.) We do not carry out this update until the focus is no longer on the text field, to avoid carrying out this (potentially expensive) computation too often.

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+                    .width(LocalConfiguration.current.screenWidthDp.dp - Padding.Large * 2)
+                    .height(300.dp),
+            ) {
+                items(filteredStations.toList()) { station ->
```
agree with this CoPilot suggestion here!

AmeliaIsCoding replied:
@lancylot2004 and Nick, did you see the comment here please?

NCHS6 replied:
FilteredList is a set not a list, so I think we need to convert it to list. This suggested change caused an error

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+) {
+    var expanded by remember { mutableStateOf(false) }
+    var searchQuery by remember { mutableStateOf("") }
+    val stations by Client.stations.collectAsState()
```
is it ideal to call the API twice for both dropdowns? how could we improve this?

NCHS6 replied:
My understanding of this code is that when the Client is initialised it grabs all the station names and stores them in a flow. The drop down menu then only collect the stations from the flow. The API is only called once I believe

AmeliaIsCoding replied:
nice!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+    toStation: Station?,
+    uriHandler: UriHandler,
+) {
+    requireNotNull(fromStation) { "Origin station not selected!" }
```
nice!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/theme/Const.kt:
```diff
@@ -0,0 +1,18 @@
+object Padding {
+    val Small = 8.dp
+
+    val Large = 32.dp
```
in my experience, I'm not sure how useful it would be to store these values as constant vals. You're likely to need quite a range of padding values for even simple designs, which could get out of hand if you store them in here. Also, it unnecessarily couples different components - if you need to change the padding on one component, you shouldn't necessarily need to change it on another. In my experience it's better to hardcode. What do you think?

lancylot2004 replied:
Point taken! The goal definitely wasn't to encode all possible UI sizing constants, just ones that might be used a lot and are stylistically consistent across the app. Happy to magic number everything; what do you prefer?

AmeliaIsCoding replied:
definitely prefer using numbers directly, in my experience!

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun StationSelect(
```
it's useful and conventional to describe the component using a noun that describes the main type of component. so `StationDropdown` might be good here

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+        ) {
+            LazyColumn(
+                Modifier
+                    .width(LocalConfiguration.current.screenWidthDp.dp - Padding.Large * 2)
```
this uses the width of the device screen rather than the parent component, resulting in the possibility of this component overflowing. if you wanted to change the parent component width you might have to update this, which isn't very maintainable. can you come up with an alternative (eg refer to the parent width - `Modifier.matchParentSize()` may be useful)

lancylot2004 replied:
I do believe this is the bug I was mentioning earlier. I've put in a solution now that uses a `.onGloballyPositioned` on the text field to track width - not the prettiest, but as you point out better than the potential fragility of relying on screen sizes!

`.matchParentSize()` afaik is only available in a `BoxScope` (the `ExposedDropdownMenuBoxScope` variant of which does not inherit from), and not explicitly specifying widths or relying on intrisic measurements would result in...

```
Asking for intrinsic measurements of SubcomposeLayout layouts is not supported. This includes components that are built on top of SubcomposeLayout, such as lazy lists, BoxWithConstraints, TabRow, etc. To mitigate this:                                                                                           
- if intrinsic measurements are used to achieve 'match parent' sizing, consider replacing the parent of the component with a custom layout which controls the order in which children are measured, making intrinsic measurement not needed                                                                                               
- adding a size modifier to the component, in order to fast return the queried intrinsic measurement.
```

And a fatal crash...

AmeliaIsCoding replied:
fair enough! happy for you to keep - maybe worth chatting to Clara and Akshay as I think they came up against the same thing and also had the `.onGloballyPositioned` solution

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+    val stations by Client.stations.collectAsState()
+    val focusRequester = remember { FocusRequester() }
+
+    val filteredStations = remember(searchQuery, stations) {
```
this will recalculate a lot as a user types but might not necessarily change.
Consider using `derivedStateOf`. This updates whenever the inputs change, as required, but only recomposes the view if the result changes. This could improve performance :)

AmeliaIsCoding replied:
https://medium.com/androiddevelopers/jetpack-compose-when-should-i-use-derivedstateof-63ce7954c11b

lancylot2004 replied:
Changes made :) Thank you for catching this

AmeliaIsCoding commented on app/src/main/kotlin/com/example/trainboard/ui/RootView.kt:
```diff
@@ -0,0 +1,193 @@
+    val stations by Client.stations.collectAsState()
+    val focusRequester = remember { FocusRequester() }
+
+    val filteredStations = remember(searchQuery, stations) {
```
should we only show stations with a non-null CRS, if these are the only ones that the user can perform a search with?

lancylot2004 replied:
Yes! This is indeed the (unfortunately implicit) behaviour right now. How should we make it more clear?

AmeliaIsCoding replied:
`handleSearch` has validation for null CRS's, which imply you allow the user to select a station with a null CRS.
maybe we should have 1 type for the return station type, and another type for the filtered station with a non-null CRS!

 we can also make this more clear with our variable names. I think in this class its okay to refer to stations as `stations`, since here we only care about stations with non null CRS. but in client, you could distinguish between these two cases with your variable names. 
I'm imagining something like
`val stations by Client.stationsWithNonNullCrs.collectAsState()`

AmeliaIsCoding reviewed (commented):
Great work! Nice work implementing searching for the station, focusing, and fetching the whole list of stations. 
Left a few comments for you to have a look at :)

lancylot2004 committed:
fix: Minor fixes according to #1.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

AmeliaIsCoding commented:
Also have you thought about adding previews to your Jetpack Compose files? https://developer.android.com/develop/ui/compose/tooling/previews

lancylot2004 commented:
> Also have you thought about adding previews to your Jetpack Compose files? [developer.android.com/develop/ui/compose/tooling/previews](https://developer.android.com/develop/ui/compose/tooling/previews)

We should! Apologies I am used to KMP dev where previews are not possible in common code...

lancylot2004 committed:
fix: Basic error handling with requests.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

# Conflicts:
#	app/src/main/kotlin/com/example/trainboard/api/Client.kt
#	app/src/main/kotlin/com/example/trainboard/ui/RootView.kt
#	app/src/main/kotlin/com/example/trainboard/ui/StationDropdown.kt

lancylot2004 committed:
fix: API errors no longer causes a crash; proper solution in #2.

Co-authored-by: Nickleby Simmonds <nickleby.simmonds@softwire.com>

AmeliaIsCoding commented:
Nice! outstanding issues:
- using padding values directly rather than variables
- more obviously using only stations with non-null CRS's
then i'll be happy to approve!

lancylot2004 committed:
fix: Removes `Padding` constant values per comments on #1.

lancylot2004 committed:
fix: Explicitly only deserialise stations with CRSes per comments on #1.

lancylot2004 commented:
Fixed!

For non-null CRSes, I've written a custom (de)serialiser which ignores stations with null CRSes. Hopefully this is a clean and accepted solution in this case!
